module/zfs/dmu.c

```diff
@@ -768,6 +769,30 @@ dmu_objset_zfs_unmounting(objset_t *os)
 	return (B_FALSE);
 }
 
+/*
+ * Return the amount of data contained in the given dnode's chunk
+ */
+static uint64_t
+dmu_dnode_count_data_in_chunk(dnode_t *dn, uint64_t offset, uint64_t chunk_len)
+{
+	uint64_t holes_size = 0;
+	uint64_t orig_offset = offset, last_offset = offset;
+
+	while (offset < orig_offset + chunk_len) {
+		if (dnode_next_offset(dn, 0, &offset, 1, 1, 0) != 0)
+			return (chunk_len - holes_size);
```

Comment:
I think this means that we'll read all the L1 indirect blocks.  I think that should be OK since we're already going to read them in dmu_free_long_range_impl() since we call dmu_tx_hold_free()?

Reply:
Yes I figure we can do this a chunk at a time since the chunks will be getting accessed anyway. I'm going to look into the testing hangs this is causing soon btw